### PR TITLE
Create migrations and models for Faq Sections and Subsections

### PR DESCRIPTION
**Pull request by davidbeig**

### CODE REVIEW

File: db/migrations/20220413104717_goteo_new_faq.php
Comment (carlos-mg89) [inline db/migrations/20220413104717_goteo_new_faq.php:26]: This UPDATE query looks important to me.

If by any chance, an exception is thrown, the query won't be executed and it might let the system in an inconsistent status (we might expect all FAQs to have a slug value).

Perhaps, we could write a log that indicates that there was an issue with this UPDATE query for a given FAQ id. And deal with it in order to fix it, and let every FAQ with a slug.

File: db/migrations/20220413104717_goteo_new_faq.php
Comment (carlos-mg89) [inline db/migrations/20220413104717_goteo_new_faq.php]: Are we really going to have so many FAQ sections that we need a BIGINT for the ID?

Maybe it's enough with a simple INT(11) type.

The same applies to the other tables being created in this migration.

File: src/Goteo/Model/Faq.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq.php]: We could remove the comment

File: src/Goteo/Model/Faq.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq.php:35]: Why is it necessary to have a fallback to the method `getById`?

File: src/Goteo/Model/Faq.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq.php]: The `subsection_id` column, is actually a `BIGINT`.

`ALTER TABLE faq ADD COLUMN subsection_id BIGINT(20) UNSIGNED NOT NULL AFTER title;`

If we allow the `$subsection` parameter to be a string, or anything that's not an `INT`, it will always return an empty list. Or even worse, the query might crash / return an exception.

I propose to set the `$subsection` parameter, as an `INT`, and remove the default `node` value.

File: src/Goteo/Model/Faq.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq.php]: This comment can be removed

File: src/Goteo/Model/Faq.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq.php]: Could we remove the `$count` param from the `getList()` method, and create a `getListCount()` method, for example, so `getList()` is more concise?

File: src/Goteo/Model/Faq.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq.php]: Can we remove this line please?

File: src/Goteo/Model/Faq.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq.php]: Can we remove this line please?

File: src/Goteo/Model/Faq/FaqSection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSection.php]: Can we remove this comment please?

File: src/Goteo/Model/Faq/FaqSection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSection.php]: Indentation doesn't follow the 4 spaces used in the rest of the file

File: src/Goteo/Model/Faq/FaqSection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSection.php]: Can we remove this comment please?

File: src/Goteo/Model/Faq/FaqSection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSection.php]: Could we separate this in a different method? `getListCount()` for example.

File: src/Goteo/Model/Faq/FaqSection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSection.php]: Can we remove this comment please?

File: src/Goteo/Model/Faq/FaqSection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSection.php]: Could we remove the empty lines?

File: src/Goteo/Model/Faq/FaqSection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSection.php]: Shouldn't the method name be `getFaqBySection`?

File: src/Goteo/Model/Faq/FaqSubsection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSubsection.php:82]: Could we separate these lines in a new method that only deals with returning the count?

File: src/Goteo/Model/Faq/FaqSubsection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSubsection.php:94]: Can we remove this comment please?

File: src/Goteo/Model/Faq/FaqSubsection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSubsection.php]: Should the name be `getFaqBySubsection`?

File: src/Goteo/Model/Faq/FaqSubsection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSubsection.php]: Let's add array and bool as parameter type and return type instead. Won't you agree @davidbeig ?

File: src/Goteo/Model/Faq/FaqSubsection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSubsection.php]: Let's add the parameter and return types instead of comments

File: tests/Goteo/Model/FaqTest.php
Comment (carlos-mg89) [inline tests/Goteo/Model/FaqTest.php]: If there's an error while removing the FaqSection, the test should simply fail.

There's no need to pollute the PHPUnit logs. Right now the shows the message between the test execution and is not clear what's going on.

Tests shouldn't print anything.

**Review by carlos-mg89 — CHANGES_REQUESTED** [review]
A few comments have been left. Requires more work before approval.

File: db/migrations/20220413104717_goteo_new_faq.php
Comment (davidbeig) [reply db/migrations/20220413104717_goteo_new_faq.php:26]: @javicarrillo I think it was you who did this. I don't think the library we are using rolls back if the postup function does not work... But yes, at least we should log the errors.

File: db/migrations/20220413104717_goteo_new_faq.php
Comment (davidbeig) [reply db/migrations/20220413104717_goteo_new_faq.php]: No, we won't. You are right

File: db/migrations/20220413104717_goteo_new_faq.php
Comment (javicarrillo) [reply db/migrations/20220413104717_goteo_new_faq.php:26]: I think it isn't critical because we have in Faq model cover the case "slug empty". I would add log in a future, right now there isn't logs in migrations and it would be an other rework task.

File: src/Goteo/Model/Faq.php
Comment (davidbeig) [reply src/Goteo/Model/Faq.php:35]: I think it was meant to be used with the slug property, but since some faqs might not have it, we could ask directly with the id if slug is not set. Obviously we could ask first the getBySlug and if there is no faq then ask with the getById()...

File: src/Goteo/Model/Faq/FaqSection.php
Comment (davidbeig) [reply src/Goteo/Model/Faq/FaqSection.php]: Since it's a method inside FaqSection i think we can just call it getFaqs. Thanks for pointing it out.

File: src/Goteo/Model/Faq/FaqSection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSection.php]: Great refactor @davidbeig !

File: src/Goteo/Model/Faq/FaqSection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSection.php:115]: The variable `$filter` will always be empty (since it has just been defined as empty 2 lines above), so this IF clause is never going to happen.

The same happens in the `getList()` method. The `$filter` variable is always an empty array.

Either we should be checking for `$filters` (the method parameter), or this IF could just be removed.

I think that issue is that we should be checking if `$filters` has any item inside instead of `$filter`.

File: src/Goteo/Model/Faq/FaqSection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSection.php]: @davidbeig The parameter `$limit` is not being used inside the method. Maybe we could just remove it?

File: src/Goteo/Model/Faq/FaqSection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSection.php:159]: @davidbeig I propose you to move the `$fields` variable of this method, into a class constant. With a name that highlights the purpose of properties being saved.

This way, the method is much smaller, and perhaps the constant can be reused somewhere else.

Something like:

```
const EDITABLE_PROPERTIES = [
            'id',
            'name',
            'slug',
            'icon',
            'banner_header',
            'button_action',
            'button_url',
            'lang',
            'order'
        ];
```

File: src/Goteo/Model/Faq/FaqSubsection.php
Comment (carlos-mg89) [inline src/Goteo/Model/Faq/FaqSubsection.php:143]: @davidbeig maybe you would like to move this variable as a class constant too. As I proposed you in `FaqSection` class.

File: tests/Goteo/Model/FaqTest.php
Comment (carlos-mg89) [inline tests/Goteo/Model/FaqTest.php:167]: Perfect!

File: src/Goteo/Model/Faq/FaqSection.php
Comment (davidbeig) [reply src/Goteo/Model/Faq/FaqSection.php:115]: The `$filter` variable is to be filled by conditions if the $filters array has some data to be filtered with. Most `getList` functions inside our models work this way. It's true that we can just have this getList without filters and that it only returns data based on `$offset` and `$limit`

File: src/Goteo/Model/Faq/FaqSection.php
Comment (davidbeig) [reply src/Goteo/Model/Faq/FaqSection.php:159]: You are right, but for now i would just keep this way since it's the one done in most models. After all the necessary reworks we'll be switching to Doctrine and then we won't need this kind of functions.

**Comment by davidbeig** [comment]
I have changed already all your requests.

**Comment by carlos-mg89** [comment]
I still see some unresolved topics, however, if you think they aren't important, I'll gladly approve the PR :) But just in case, review the unresolved conversations and write me again please.

**Review by carlos-mg89 — APPROVED** [review]
It looks pretty good @davidbeig ! Good job!

**Comment by davidbeig** [comment]
This PR is not meant to be merged. It will be superseeded by a new PR with all the migrations, admin and view of the new FAQS

**Pull request closed without merging**
